### PR TITLE
Add failover APIs in policy

### PR DESCRIPTION
**Pull request by Poor12**
**What type of PR is this?**
/kind api-change

**What this PR does / why we need it**:
Add failover APIs according to #3133.

**Which issue(s) this PR fixes**:
Fixes #

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:
None

### CODE REVIEW

File: pkg/apis/work/v1alpha2/binding_types.go [hunk 1]
```diff
@@ -298,21 +298,6 @@ type ResourceBindingList struct {
 	Items []ResourceBinding `json:"items"`
 }
 
-// ResourceHealth represents that the health status of the reference resource.
-type ResourceHealth string
```
Comment (XiShanYongYe-Chang) [inline pkg/apis/work/v1alpha2/binding_types.go:302]: Does the deletion affect published APIs?

File: pkg/apis/policy/v1alpha1/propagation_types.go [hunk 2]
```diff
@@ -143,6 +148,128 @@ type FieldSelector struct {
...
+	HealthyState ResourceHealth `json:"healthyState,omitempty"`
+}
+
+// DecisionConditions represents the decision conditions of performing the failover process.
+type DecisionConditions struct {
+	// TolerationSeconds represents the period of time Karmada should wait
+	// after reaching the desired state before performing failover process.
+	// If not specified, Karmada will immediately perform failover process.
+	// Defaults to 10s.
+	// +kubebuilder:default=10
```
Comment (RainbowMango) [inline pkg/apis/policy/v1alpha1/propagation_types.go:259]: Defaults to 300s.

File: pkg/apis/policy/v1alpha1/propagation_types.go [hunk 3]
```diff
@@ -143,6 +148,128 @@ type FieldSelector struct {
...
+	// +optional
+	TolerationSeconds *int32 `json:"tolerationSeconds,omitempty"`
+
+	// HealthyState refers to the healthy status reported by the Karmada resource
+	// interpreter.
+	// Valid options are "Unhealthy", "Unknown", or both.
```
Comment (RainbowMango) [inline pkg/apis/policy/v1alpha1/propagation_types.go:265]: keep only one option `Unhealthy`.

File: pkg/apis/policy/v1alpha1/propagation_types.go [hunk 4]
```diff
@@ -143,6 +148,128 @@ type FieldSelector struct {
...
+	// When the application reaches the desired HealthyState, Karmada will perform failover process
+	// after TolerationSeconds is reached.
+	// +kubebuilder:validation:MinItems=1
+	// +required
+	HealthyState []ResourceHealth `json:"healthyState,omitempty"`
```
Comment (RainbowMango) [inline pkg/apis/policy/v1alpha1/propagation_types.go:270]: 1. []ResourceHealth --> ResourceHealth
2. remove omitempty

Comment (Poor12) [reply pkg/apis/policy/v1alpha1/propagation_types.go]: Updated.

File: pkg/apis/policy/v1alpha1/propagation_types.go [hunk 5]
```diff
@@ -143,6 +148,128 @@ type FieldSelector struct {
...
+	// after reaching the desired state before performing failover process.
+	// If not specified, Karmada will immediately perform failover process.
+	// Defaults to 300s.
+	// +kubebuilder:default=300
+	// +optional
+	TolerationSeconds *int32 `json:"tolerationSeconds,omitempty"`
+
+	// HealthyState refers to the healthy status reported by the Karmada resource
+	// interpreter.
+	// Valid options are "Unhealthy".
```
Comment (RainbowMango) [inline pkg/apis/policy/v1alpha1/propagation_types.go:265]:
```suggestion
	// Valid option is "Unhealthy".
```

File: pkg/apis/policy/v1alpha1/propagation_types.go [hunk 6]
```diff
@@ -143,6 +148,128 @@ type FieldSelector struct {
...
+	// DelaySeconds refers to a period of time after the control plane collects
+	// the status of the application for the first time.
+	// If specified, the failover process will be started after DelaySeconds is reached.
+	// It can be used simultaneously with HealthyState and does not affect each other.
+	// +optional
+	DelaySeconds *int32 `json:"delaySeconds,omitempty"`
+
+	// HealthyState refers to the healthy status reported by the Karmada resource
+	// interpreter.
+	// Valid options are "Healthy".
```
Comment (RainbowMango) [inline pkg/apis/policy/v1alpha1/propagation_types.go:246]:
```suggestion
	// Valid option is "Healthy".
```

Comment (Poor12) [reply pkg/apis/policy/v1alpha1/propagation_types.go]: Done.

**Review by RainbowMango — APPROVED** [review]
/lgtm
/approve

**Pull request merged**